db/migrate/20200728054557_refer_config_proxy_to_http_proxy.rb

```diff
@@ -0,0 +1,59 @@
+class ReferConfigProxyToHttpProxy < ActiveRecord::Migration[6.0]
+  def up
+    add_column :foreman_virt_who_configure_configs, :http_proxy_id, :integer
+
+    if User.where(login: User::ANONYMOUS_ADMIN).exists?
+      User.as_anonymous_admin do
+        @http_proxies = HttpProxy.all
+        ::ForemanVirtWhoConfigure::Config.find_each do |config|
+          if config.proxy.present?
+            create_and_refer_http_proxy_if_not_exist(config)
+          end
+        end
+      end
+    end
+
+    remove_column :foreman_virt_who_configure_configs, :proxy
+  end
+
+  def down
+    add_column :foreman_virt_who_configure_configs, :proxy, :string
+
+    if User.where(login: User::ANONYMOUS_ADMIN).exists?
+      User.as_anonymous_admin do
+        ::ForemanVirtWhoConfigure::Config.find_each do |config|
+          if config.http_proxy_id.present?
+            config.proxy = config.http_proxy.url
+            config.save!
+          end
+        end
+      end
+    end
+
+    remove_column :foreman_virt_who_configure_configs, :http_proxy_id
+  end
+
+  private
+
+  def create_and_refer_http_proxy_if_not_exist(config)
```

Comment:
There's quite a bit of behavior for this method, and I'd feel more comfortable with it if we had some unit tests in order to verify a handful of cases like when the HttpProxy already exists, when it does not exist, and various permutations of the current proxy string from an ip address, to http://proxy.example.com, with port, without port, with auth (http://user:pass@proxy.example.com:8181) and so on.

The easiest way to test might be to make this a class method and require this file in your test class and call it: `ReferConfigProxyToHttpProxy.migrate_config(http_proxies, config)` for example. Easy to test then, so try your hardest to break it :)